### ISSUE TITLE
Page-specific data leaking to next page when route unchanged (starting in v0.12).

### ISSUE THREAD

**Issue by morleytatro** [sveltejs/sapper#264]
With the most recent release, the route's page component doesn't get torn down and rebuilt if the route stays the same but the params and/or query changes. This causes undesirable behavior with data stored outside of the main object returned from preload. For instance, say you have a product page that sets a product object into local state from its preload method. If the product has variations and you make an AJAX call inside oncreate to get those variations, you could store the returned variations inside the main product object, but this would require preserving the product's current properties either through object rest spread or Object.assign.

```
const { product } = this.get();
this.set({
  product: {
    ...product,
    variations
  }
});
```

```
const { product } = this.get();
this.set({
  product: Object.assign(
    {},
    product,
    variations
  )
});
```

Ideally, you could simply save the variations simply using ```this.set({ variations })```, but with the current architecture that would leave the variations data in place if someone navigates to a different product from the page. It seems inefficient to have to clean up on new "page loads" and also not ideal to move functionality from oncreate to preload to ensure it always gets run (a separate workaround for the non-teardown behavior).

**Comment by Rich-Harris** [comment]
This is most likely fixed in 0.15 — Sapper will destroy and recreate components as necessary to ensure that `preload` and `oncreate` run when they're supposed to. If you get a chance to upgrade could you let me know if it's resolved? Thanks

**Comment by morleytatro** [comment]
Looks good Rich. So far so good!
